docs/docsite/rst/dev_guidelines.rst

```diff
@@ -497,6 +497,79 @@ E.g. if boto3 returns a value called 'SecretAccessKey' do not change it to 'Acce
 
 .. _ansible_collections.amazon.aws.docsite.dev_policies:
 
+Info modules
+------------
+
+Info modules that can return information on multiple resources should return a singular value representing a list of dictionaries,
+with each dictionary containing information about that particular resource (i.e. ``security_groups`` in ``ec2_group_info``).
+In cases where the _info module only returns information on a singular resource (i.e. ``ec2_tag_info``),
+a singular dictionary should be returned as opposed to a list of dictionaries.
+In cases where the _info module returns no instances, and empty list '[]' should be returned.
+Keys in the returned dictionaries should follow the guidelines above and use snake_case.
+If a return value can be used as a parameter for its corresponding main module, the key should match either
+the parameter name itself, or an alias of that parameter. The following is an example of improper usage of an info module with its respective main module:
+
+.. code-block:: yaml
+
+    "security_groups": [
+        {
+            "description": "Created by ansible integration tests",
+            "group_id": "sg-050dba5c3520cba71",
+            "group_name": "ansible-test-87988625-unknown5c5f67f3ad09-icmp-1",
+            "ip_permissions": [
+                {
+                    "from_port": 3,
+                    "ip_protocol": "icmp",
+                    "ip_ranges": [
+                        {
+                            "cidr_ip": "172.16.40.10/32"
+                        },
+                        {
+                            "cidr_ip": "10.0.0.0/8"
+                        }
+                    ],
+                    "ipv6_ranges": [],
+                    "prefix_list_ids": [],
+                    "to_port": 8,
+                    "user_id_group_pairs": []
+                }
+            ],
+            "ip_permissions_egress": [
+                {
+                    "ip_protocol": "-1",
+                    "ip_ranges": [
+                        {
+                            "cidr_ip": "0.0.0.0/0"
+                        }
+                    ],
+                    "ipv6_ranges": [],
+                    "prefix_list_ids": [],
+                    "user_id_group_pairs": []
+                }
+            ],
+            "owner_id": "721066863947",
+            "tags": {},
+            "vpc_id": "vpc-0cbc2380a326b8a0d"
+        }
+    ]
+
+The output above is what was returned by the ``ec2_group_info`` module. It correctly returns a list of dictionaries called ``security_groups``; however,
+the parameter names do not match those of the ``ec2_group`` module. ``ec2_group`` requires ``name``, meanwhile this returns ``group_name``.
```

Comment:
It's generally a good idea to follow a "negative" example with a "positive" example, and make sure that it's clear one example is "bad" the other is "good" and why.  If someone wants to come back to an example it's easier to follow a "good" example than to try and compare a "bad" example to the written text and try to figure out what was wanted.

I'd suggest keeping your existing example, maybe adding a boto3 style tag list, and then adding a "corrected" example underneath.

Reply:
Also, some modules return a list of dictionaries called `result`. I guess we should also uniform this across the modules and it would be also easier for the structure of the result and also for the user instead of having sometimes  `security_groups`, sometimes `result`, or `clusters` or whatever.

Reply:
> Also, some modules return a list of dictionaries called `result`.

Yeah, let's define a 'standard', for not just the _info modules, but also the main modules themselves.  I suspect it'll be a pretty substantial cleanup effort though, but at least if we document what things should be we stop building the technical debt.

Reply:
@alinabuzachis what about something like `instances` or `resources` to name the list returned? 
@tremble i added a task box to the summary of things to address - I think it makes sense to have the main modules just return their keys as is without wrapping them in a dict. Thoughts?

Reply:
Personally I prefer when things are wrapped: it gives you a clean namespace to work in and it's clear what's the scope of the object, versus something that Ansible injected (for example "changed", "invocation", or "resource_actions" from the callback plugin).  Additionally, having a distinct dictionary that's returned is closer to the behaviour of our _info modules.

Reply:
That makes sense to me. Curious what others think @alinabuzachis @jillr @markuman

Reply:
> Personally I prefer when things are wrapped: it gives you a clean namespace to work in and it's clear what's the scope of the object, versus something that Ansible injected (for example "changed", "invocation", or "resource_actions" from the callback plugin). Additionally, having a distinct dictionary that's returned is closer to the behaviour of our _info modules.

this makes sense. Having something to wrap the result/output that module returns in would be a good idea I suppose.
For example
```
{
    "changed": false,
    "failed": false,
    "result": [ #variable name to be decided that could be used commonly across all/most of the modules
                    {
                    ...
                    },
                    #this is the result/output returned by the module task
                    {
                    ...
                    },
                    .
                    .
                  ]
}


```

Reply:
> That makes sense to me. Curious what others think @alinabuzachis @jillr @markuman @mandar242

@jatorcasso I agree with @tremble. How we'd like to call that returned dictionary, simply `result`?

Reply:
`result` is fine, but to me that feels like `result` would be what we registered. For example:
```
  ec2_eni:
    eni_id: "{{ eni_id_1 }}"
    name: "{{ resource_prefix }}"
    subnet_id: "{{ vpc_subnet_id }}"
    state: absent
  register: result
```

to get the eni details: `result.result`

what about the module name? so in the case above:
to get the eni details: `result.ec2_eni`

to me that feels more practical

Reply:
Not all the people call the registered result as ‘result’. It’s up to them how calling it. Using the module’s name might be an alternative, but not sure if it brings more complexity (more text to be written) to access the result and probably not so immediate. BTW, it works for me in either way. Let’s see what @tremble and @jillr suggest.

Reply:
> Not all the people call the registered result as ‘result’. BTW, it’s up to them how calling it. 

`register: out` is also widespread :smiley: 

> what about the module name? so in the case above:  
> to get the eni details: result.ec2_eni

I like that approach.

Reply:
> what about the module name? so in the case above: to get the eni details: `result.ec2_eni`

Either the module name, or the name of the object that boto3 returns, for example `result.network_interface`:

```
{
    "changed": false,
    "failed": false,
    "network_interface": {
        network_interface_id: "eni-123456789abcd",
        ...
    },
    "response_metadata": {
    }
}
```

Occasionally it's helpful to also return the 'response metadata'.  I'd say that belongs back at the top level, but all under response_metadata

Reply:
> to get the eni details: `result.result`

I'd beware of `result`, it's a little too close to `results` from loops:

`result.results[0].result.network_interface_id`

Reply:
> Either the module name, or the name of the object that boto3 returns, for example `result.network_interface`:

This could also work, but I feel that adds an additional layer of complexity where the user has to reference our docs to see what that name actually is. I think it should be up to us to rename it accordingly to retain consistency in the collection. Whether that's the same key name (`result`), or the module name (`ec2_eni`), at least as a user you will know whats being returned without having to spend too much effort. 

> Occasionally it's helpful to also return the 'response metadata'. I'd say that belongs back at the top level, but all under response_metadata

I agree with response_metadata being at the top level, to distinguish itself from the actual resource data (i.e. network_interface)

Reply:
> > what about the module name? so in the case above:
> > to get the eni details: result.ec2_eni
> 
> I like that approach.

The only part of this approach I _dont_ like is regarding info modules. For example, lambda_info module. `lambda_info` would be used to represent the list of lambdas returned, with each `lambda_info[i]` representing info of lambda x. Something like `lambdas` would probably be a better name, but I'm not sure the best approach for naming the info modules return data. Maybe `lambda_info` is fine.

Reply:
> Occasionally it's helpful to also return the 'response metadata'. I'd say that belongs back at the top level, but all under response_metadata

Same for info modules as well? so, `lambda_info` returns info on all lambdas, then `response_metadata` would return the metadata. Or maybe each item in the list `lambda_info` should contain both, which in that case, we could use the name of the main module to wrap its details, and `response_metadata` would be included in each `lambda_info` item. That would look like this:

```
response_metadata for each lambda returned in lambda info
"lambda_info": [
   "lambda": {
      "function_name": abc
      ...
   }
   "response_metadata": {
      "http_headers": ...
   },
   ...
]
```
versus
```
response_metadata for entire lambda_info call
"lambda_info": [
   {
      "function_name": abc
      ...
   },
   ...
]
"response_metadata": {
   "http_headers": ...
}
```

Reply:
Looking at what we've already got, naming it after the module (eg lambda_info) is going to require touching pretty much *every* module, it's also going to result in different keys for example and example_info.  We already expect people to look at the documentation to describe the parameters for a module, I don't think it's a stretch to expect them to look at the docs for the return values (beyond changed/success/failed)

Reply:
I would prefer not to have to change a majority of modules (and user playbooks) just to set a standard.  By that same logic, I'd guess we're well past the point where we could reasonably pick a single key (like "result") to return from all modules.  Having a sensible approximation of the resource name (lambda[s], ec2_eni[s]) that can be referenced in the docs feels like the option that will be cleanest to implement without majorly negatively impacting users.

Reply:
@jillr That sounds good to me